File: cli/src/wsc/ops/docs.py
"""wsc docs — official library docs via Context7, falling back to GitHub README via Firecrawl."""

from __future__ import annotations

from typing import Any, Dict, Optional

from wsc import audit
from wsc.ops._chain import chain_failed_payload, filtered_chain, query_fingerprint, run_chain
from wsc.providers import FetchedPage, ProviderError


def run(
    library: str,
    *,
    topic: Optional[str] = None,
    version: Optional[str] = None,
    correlation_id: Optional[str] = None,
    no_receipt: bool = False,
) -> Dict[str, Any]:
    chain = filtered_chain("library_docs")

    def context7_action(provider):
        candidates = provider.resolve_library(library)
        if not candidates:
            raise ProviderError(f"context7: no library found for {library!r}")
        # Take the top candidate; library_id is the path part of the URL.
        top = candidates[0]
        # url is like https://context7.com/api/v1/<library_id>
        library_id = top.url.split("/api/v1", 1)[-1] or "/" + library
        page: FetchedPage = provider.get_docs(library_id, topic=topic)
        return {"library_id": library_id, "page": page, "candidates": candidates}

    def firecrawl_fallback(provider):
        # Best-effort: scrape the GitHub README for ``<library>``. Many libs
        # are at github.com/<library>/<library> (e.g. vercel/next.js); this is
        # a best-guess fallback only used when context7 is unavailable.
        guesses = [
            f"https://raw.githubusercontent.com/{library}/{library}/main/README.md",
            f"https://github.com/{library}/{library}",
        ]
        last_err: Optional[Exception] = None
        for url in guesses:
            try:
                page = provider.scrape(url)
                page.status = "degraded"  # readme is not the real docs
                return {"library_id": library, "page": page, "candidates": []}
            except ProviderError as exc:
                last_err = exc
                continue
        raise last_err or ProviderError("firecrawl: no readme found for fallback")

    actions = {"context7": context7_action, "firecrawl": firecrawl_fallback}

    with audit.start_call(
        "docs", provider=chain[0] if chain else None,
        correlation_id=correlation_id, no_receipt=no_receipt,
    ) as receipt:
        receipt.update(query_fingerprint(library))
        receipt["params"] = {"topic": topic, "version": version}
        active, result, fallback = run_chain(chain, actions)
        receipt["fallback_chain"] = fallback
        if active is None:
            receipt["status"] = "error"
            return chain_failed_payload("docs", fallback)
        receipt["provider"] = active
        page: FetchedPage = result["page"]
        receipt["selected_urls"] = [page.url] if page.url else []
        receipt["selected_count"] = 1
        receipt["results_count"] = 1
        if page.status == "degraded":
            receipt["status"] = "degraded"
        return {
            "ok": True,
            "operation": "docs",
            "provider": active,
            "library": library,
            "library_id": result["library_id"],
            "topic": topic,
            "page": page.to_dict(),
            "fallback_chain": fallback,
            "status": page.status,
            "returncode": 0,
        }

File: src/ops/docs.ts
/** wsc docs — Context7 primary, Firecrawl GitHub README fallback. */

import { withCall } from "../audit.js";
import { chainFailedPayload, filteredChain, queryFingerprint, runChain } from "./_chain.js";
import type { Action } from "./_chain.js";
import { Context7Provider, FetchedPage, FirecrawlProvider, ProviderError } from "../providers/index.js";

export interface DocsOptions {
  topic?: string;
  version?: string;
  correlationId?: string;
  noReceipt?: boolean;
}

export interface DocsPayload {
  ok: boolean;
  operation: "docs";
  provider: string | null;
  library: string;
  library_id?: string;
  topic?: string | null;
  page?: ReturnType<FetchedPage["toJSON"]>;
  fallback_chain: ReturnType<typeof filteredChain> extends readonly (infer _)[] ? Array<{ from: string; reason: string; error?: string }> : never;
  status?: string;
  returncode: number;
  error?: string;
}

export async function run(library: string, opts: DocsOptions = {}): Promise<DocsPayload> {
  const chain = filteredChain("library_docs");

  const context7Action: Action<{ library_id: string; page: FetchedPage }> = async (provider) => {
    const ctx = provider as Context7Provider;
    const candidates = await ctx.resolveLibrary(library);
    if (candidates.length === 0) throw new ProviderError(`context7: no library found for ${JSON.stringify(library)}`);
    const top = candidates[0]!;
    const libraryId = (top.url.split("/api/v1", 2)[1] ?? "/" + library) || "/" + library;
    const page = await ctx.getDocs(libraryId, { topic: opts.topic });
    return { library_id: libraryId, page };
  };

  const firecrawlAction: Action<{ library_id: string; page: FetchedPage }> = async (provider) => {
    const fc = provider as FirecrawlProvider;
    const guesses = [
      `https://raw.githubusercontent.com/${library}/${library}/main/README.md`,
      `https://github.com/${library}/${library}`,
    ];
    let lastErr: unknown = null;
    for (const url of guesses) {
      try {
        const page = await fc.scrape(url);
        page.status = "degraded";
        return { library_id: library, page };
      } catch (err) {
        lastErr = err;
      }
    }
    throw lastErr instanceof Error ? lastErr : new ProviderError("firecrawl: no readme found for fallback");
  };

  return await withCall(
    "docs",
    { provider: chain[0] ?? null, correlationId: opts.correlationId, noReceipt: opts.noReceipt },
    async (receipt) => {
      Object.assign(receipt, queryFingerprint(library));
      receipt.params = { topic: opts.topic ?? null, version: opts.version ?? null };
      const { active, result, fallback } = await runChain(chain, {
        context7: context7Action,
        firecrawl: firecrawlAction,
      });
      receipt.fallback_chain = fallback;
      if (active === null || result === null) {
        receipt.status = "error";
        const failed = await chainFailedPayload("docs", fallback);
        return {
          ...failed,
          library,
          fallback_chain: fallback,
        } as unknown as DocsPayload;
      }
      receipt.provider = active;
      const page = result.page;
      receipt.selected_urls = page.url ? [page.url] : [];
      receipt.selected_count = page.url ? 1 : 0;
      receipt.results_count = 1;
      if (page.status === "degraded") receipt.status = "degraded";
      return {
        ok: true,
        operation: "docs",
        provider: active,
        library,
        library_id: result.library_id,
        topic: opts.topic ?? null,
        page: page.toJSON(),
        fallback_chain: fallback,
        status: page.status,
        returncode: 0,
      } as DocsPayload;
    },
  );
}
